src/realm/collection.hpp

```diff
@@ -364,12 +382,37 @@ class CollectionBaseImpl : public Interface, protected ArrayParent {
         return false;
     }
 
+    void set_owner(const Obj& obj, CollectionParent::Index index) override
+    {
+        m_obj_mem = obj;
+        m_parent = &m_obj_mem;
+        m_index = index;
+        if (obj) {
+            m_alloc = &obj.get_alloc();
+        }
+    }
+
+    void set_owner(std::shared_ptr<CollectionParent> parent, CollectionParent::Index index) override
+    {
+        m_obj_mem = parent->get_object();
+        m_col_parent = std::move(parent);
+        m_parent = m_col_parent.get();
+        m_index = index;
+        if (m_obj_mem) {
+            m_alloc = &m_obj_mem.get_alloc();
+        }
+    }
+
     using Interface::get_owner_key;
     using Interface::get_table;
     using Interface::get_target_table;
 
 protected:
-    Obj m_obj;
+    Obj m_obj_mem;
+    std::shared_ptr<CollectionParent> m_col_parent;
+    CollectionParent* m_parent = nullptr;
```

Comment:
Mad the attributes private. Maybe we are checking too much now, but better safe than sorry. Refactored the access functions so that it looks a little bit nicer..